### ISSUE TITLE
elkserver certbot fails with "connection refused" 

### ISSUE THREAD

**Issue by drfinkus** [outflanknl/RedELK#149]
I have manually modified the `mounts/redelk-config/etc/redelk/config.json` to include the correct domain and a valid email address for Let's Encrypt. However, the certbot docker image fails with "connection refused".

**Comment by fastlorenzo** [comment]
@drfinkus could you post the detailed logs? 
We've fixed several issues regarding Let's Encrypt in PR #117, you can maybe try that branch and let us know if this error still occurs.

**Comment by drfinkus** [comment]
@fastlorenzo in that branch, letsencrypt is turned off by default in config.json. Should I enable it manually, add the domain and email, and test?

**Comment by fastlorenzo** [comment]
Yes, if you want to use it

**Comment by drfinkus** [comment]
@fastlorenzo thanks, I tested it now. Unfortunately PR #117 fails with the same error, please see detailed logs below. This is the full sequence on a fresh Ubuntu 18.04 install. 

Please note that `DOMAIN.TLD` is not the actual value I ran the script with, I just edited the real domain out when pasting the logs here.

```
root@test:~# apt-get update
[...]
Fetched 18.7 MB in 3s (5709 kB/s)
Reading package lists... Done
root@test:~# apt-get install -y tmux mosh git curl gnupg jq docker docker-compose
Reading package lists... Done
Building dependency tree
Reading state information... Done
[...]
root@test:~# git clone https://github.com/outflanknl/RedELK.git /opt/RedELK
Cloning into '/opt/RedELK'...
remote: Enumerating objects: 896, done.
remote: Counting objects: 100% (896/896), done.
remote: Compressing objects: 100% (328/328), done.
remote: Total 4572 (delta 526), reused 827 (delta 502), pack-reused 3676
Receiving objects: 100% (4572/4572), 42.54 MiB | 22.04 MiB/s, done.
Resolving deltas: 100% (2822/2822), done.
root@test:~# cd /opt/RedELK
root@test:/opt/RedELK# git fetch origin pull/117/head:pr117
From https://github.com/outflanknl/RedELK
 * [new ref]         refs/pull/117/head -> pr117
root@test:/opt/RedELK# git checkout pr117
Switched to branch 'pr117'
root@test:/opt/RedELK# echo master > VERSION
root@test:/opt/RedELK# sed -i 's/MODIFYME, for example //g' /opt/RedELK/certs/config.cnf
root@test:/opt/RedELK# vi certs/config.cnf // add the correct DNS and IP values
root@test:/opt/RedELK# ./initial-setup.sh ./certs/config.cnf



    ____            _  _____  _      _  __
   |  _ \  ___   __| || ____|| |    | |/ /
   | |_) |/ _ \ / _  ||  _|  | |    | ' /
   |  _ <|  __/| (_| || |___ | |___ | . \
   |_| \__\___| \____||_____||_____||_|\_\



This script will generate necessary keys and packages for RedELK deployments



[*] Will generate TLS certificates for the following DNS names and/or IP addresses:
DNS.1 = DOMAIN.TLD
IP.1 = SANITIZED

[!] Make sure your ELK server will be reachable on these DNS names or IP addresses or your TLS setup will fail!
[*] Abort within 10 seconds to correct if needed.
[*] Creating certs dir if necessary
[*] Generating private key for CA
[*] Creating Certificate Authority
[*] Generating private key for ELK server
[*] Generating certificate for ELK server
[*] Signing certificate of ELK server with our new CA
[*] Converting ELK server private key to PKCS8 format
[*] Copying certificates to relevant redir and c2servers folders.
[*] Creating ssh directories if necessary
[*] Generating SSH key pair for scponly user
[*] Copying sshkeys to relevant folders.
[*] Copying VERSION file to subfolders.
[*] Creating TGZ packages for easy distribution


[*] Done with initial setup.
[*] Copy the redirs.tgz, c2servers.tgz and elkserver.tgz packages to every redirector, c2servers or ELK-server. Then run the relevant setup script there locally.

root@test:/opt/RedELK# cd elkserver/
root@test:/opt/RedELK/elkserver# vi mounts/redelk-config/etc/redelk/config.json // set certbot to true, add domain and email
root@test:/opt/RedELK/elkserver# ./install-elkserver.sh



    ____            _  _____  _      _  __
   |  _ \  ___   __| || ____|| |    | |/ /
   | |_) |/ _ \ / _  ||  _|  | |    | ' /
   |  _ <|  __/| (_| || |___ | |___ | . \
   |_| \__\___| \____||_____||_____||_|\_\





This script will install and configure necessary components for RedELK on ELK server



[*] Memory found available for RedELK: 15499 MB.
[*] 15-16GB memory found


[*] No 'limited' parameter found. Going for the full RedELK installation including:
- RedELK
- Jupyter notebooks
- BloodHound / Neo4j

5 Seconds to abort

[*] Starting pre installation checks
[*] Updating apt
[*] Installing apache2-utils
[*] .env file doesn't exist yet, copying from .env.tmpl
[*] Setting RedELK version in docker env file
[*] Setting kibana_system ES password
[*] Setting logstash_system ES password
[*] Setting redelk_ingest ES password
[*] Setting redelk password in elasticsearch
[*] Setting redelk password in htaccess
[*] Setting elastic ES password in Docker template
[*] Setting elastic ES password in redelk config.json
[*] Setting Kibana encryption key
[*] Adjusting memory settings for ES
[*] Adjusting memory settings for NEO4J
[*] Setting neo4j password
[*] Setting permissions on logstash configs
[*] Setting permissions on redelk logs
[*] Setting permissions on Jupyter notebook working dir
[*] Validating configured external domain name DOMAIN.TLD
[*] Domain DOMAIN.TLD seems valid. Continuing.
[*] Setting external domain name in Docker env file
[*] Setting Let's Encrypt email in Docker env file
[*] Setting CERTS_DIR_NGINX_LOCAL
[*] Setting CERTS_DIR_NGINX_CA_LOCAL
[*] Setting TLS_NGINX_CRT_PATH
[*] Setting TLS_NGINX_KEY_PATH
[*] Setting TLS_NGINX_CA_PATH
[*] Linking docker-compose.yml to the docker file used
[*] Creating password file for easy reference
[*] Copying password file for use with jupyter notebooks
[*] Running initial Let's Encrypt script
### Creating dummy certificate for DOMAIN.TLD ...
Creating network "redelk_net" with driver "bridge"
Creating volume "redelk_es_data" with local driver
Creating volume "redelk_kibana_data" with local driver
Creating volume "redelk_bloodhound_data" with local driver
Pulling nginx (nginx:latest)...
latest: Pulling from library/nginx
f7ec5a41d630: Pull complete
deac85b6e34f: Pull complete
36acc35632db: Pull complete
2e4b03180fae: Pull complete
81fc5ccdd311: Pull complete
36536268535e: Pull complete
Digest: sha256:0e9fc81d8d894c4e5d32be1b78ab6c85ca955e0ce50020e392e014f34b81c6c5
Status: Downloaded newer image for nginx:latest
Creating redelk-nginx ... done
Pulling certbot (certbot/certbot:)...
latest: Pulling from certbot/certbot
532819f3e44c: Pull complete
b4919f2c0c77: Pull complete
3c55f4db643a: Pull complete
375555f2f3fd: Pull complete
3ab006cb0616: Pull complete
802483f6ad33: Pull complete
7ffcc7425e1b: Pull complete
0249c7a566fc: Pull complete
a8b82371c50e: Pull complete
74b7ee54d393: Pull complete
eec5c40d96c8: Pull complete
96ac9ec6f167: Pull complete
8643fcc1eeea: Pull complete
Digest: sha256:e3759608b7a9b558d6391af0aa3d1f4794f6ad04a4d6e1729b4a070dd879c0ec
Status: Downloaded newer image for certbot/certbot:latest
Generating a RSA private key
................................................................................................................................++++
...................................................................................................................................................................................................................++++
writing new private key to '/etc/letsencrypt/live/DOMAIN.TLD/privkey.pem'
-----

./init-letsencrypt.sh: line 51: [: ==: unary operator expected
### Starting nginx ...
Recreating redelk-nginx ... done

### Requesting Let's Encrypt certificate for DOMAIN.TLD ...
### Removing dummy certificate folder
Starting redelk-nginx ... done
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer None
Account registered.
Requesting a certificate for DOMAIN.TLD
Performing the following challenges:
http-01 challenge for DOMAIN.TLD
Using the webroot path /var/www/certbot for all unmatched domains.
Waiting for verification...
Challenge failed for domain DOMAIN.TLD
http-01 challenge for DOMAIN.TLD
Cleaning up challenges
Some challenges have failed.

IMPORTANT NOTES:
 - The following errors were reported by the server:

   Domain: DOMAIN.TLD
   Type:   connection
   Detail: Fetching
   http://DOMAIN.TLD/.well-known/acme-challenge/ArUS9YDLrOT3Yjt8z5YrEOfBdMJoXIR9_XtIM4UF9WI:
   Connection refused

   To fix these errors, please make sure that your domain name was
   entered correctly and the DNS A/AAAA record(s) for that domain
   contain(s) the right IP address. Additionally, please check that
   your computer has a publicly routable IP address and that no
   firewalls are preventing the server from communicating with the
   client. If you're using the webroot plugin, you should also verify
   that you are serving files from the webroot path you provided.

### Reloading nginx ...
Error response from daemon: Container bf72b446982c84bfc6ac76692a7e7289142ffebdfe3ea308a9fb1cff5f440592 is restarting, wait until the container is running
[X] Error running initial Let's Encrypt script (Error Code: 1).
[*] Building RedELK from redelk-full.yml file. Docker output below.

Pulling elasticsearch (outflanknl/redelk-elasticsearch:master)...
master: Pulling from outflanknl/redelk-elasticsearch
e320e40c33f8: Pull complete
be0367056981: Pull complete
67e849725737: Pull complete
2689ab201ca3: Extracting [======================================>            ]  266.8MB/345.7MB
054ab54cebfb: Download complete
4e6ca8ea18d1: Download complete
033f0d2daef0: Download complete
3e873ca10c18: Download complete
19c868cd4818: Download complete
48c543a8b3de: Download complete
37105c49f6a3: Download complete
29b79fcef003: Download complete
15ec243ddd8b: Download complete
f9d0f15b9486: Download complete
^CERROR: Aborting.
[X] Error building RedELK using docker-compose file redelk-full.yml (Error Code: 1).
root@test:/opt/RedELK/elkserver#
```

**Comment by MarcOverIP** [comment]
@drfinkus perhaps needles question, but does the server have a public IP address? In other words, is the LE server able to reach it?

@fastlorenzo on line 51 of lets encrypt installer there is a check for a var that Im not sure is set anywhere. Either quote it or use bash specific double brackets [[..]] I dont believe it is the root of the issue here. But nonetheless should be fixed.

**Comment by drfinkus** [comment]
@MarcOverIP yes, the server has a public IP address. Also, as you can see, the error is "connection refused", which seems the indicate the server is actively refusing the connection with a RST, which implies it is reachable.

**Comment by fastlorenzo** [comment]
@drfinkus could you check with PR #152 or master?

**Comment by fastlorenzo** [comment]
> @fastlorenzo on line 51 of lets encrypt installer there is a check for a var that Im not sure is set anywhere. Either quote it or use bash specific double brackets [[..]] I dont believe it is the root of the issue here. But nonetheless should be fixed.

Fixed in #153

**Comment by drfinkus** [comment]
@fastlorenzo same certbot error in #152 and #153, `connection refused`. Were you able to reproduce this error on your end? Did #152 solve it? Again, this is all on a fresh Ubuntu 18.04 with 16GB RAM (on Digital Ocean, if that matters).

**Comment by fastlorenzo** [comment]
@drfinkus let me spin up an Ubuntu 18.04 and test it

**Comment by fastlorenzo** [comment]
ok, might have found something: it looks like nginx container cannot start as it's not able to resolve the `redelk-kibana` host (as the kibana container is not start yet at that point).
I've tried to add a dependency in the compose file for `nginx` to depend on `kibana` and `jupyter` + `bloodhound` (for the full install), but I'm hitting circular dependency issue as `jupyter` also depends on `base`, which depends on `nginx`:

https://github.com/outflanknl/RedELK/blob/d3b364558e364f484b78e7ac54087cae6243655b/elkserver/redelk-full.yml#L194-L195

@MarcOverIP / @xychix any reason why `jupyter` depends on `base`? Can we remove that dependency?

**Comment by fastlorenzo** [comment]
Next issue (by doing the previous fix on dependency change, I can get a valid certificate from LE): the "dummy" certificates in `./mounts/cerbot/conf/live/DOMAIN/*` are regular files; however, Let's Encrypt makes use of symlinks.
Due to the mounted path below, nginx can't find them:

https://github.com/outflanknl/RedELK/blob/d3b364558e364f484b78e7ac54087cae6243655b/elkserver/install-elkserver.sh#L451-L456

**Comment by fastlorenzo** [comment]
@drfinkus I've made the necessary changes in #152, please check and let me know if that works now.
@MarcOverIP  / @xychix please review the changes and merge in master whenever possible, this should be a simple fix.

```
root@test-redelk:/opt# git clone https://github.com/outflanknl/RedELK.git /opt/RedELK
Cloning into '/opt/RedELK'...
remote: Enumerating objects: 4576, done.
remote: Counting objects: 100% (900/900), done.
remote: Compressing objects: 100% (327/327), done.
remote: Total 4576 (delta 527), reused 837 (delta 507), pack-reused 3676
Receiving objects: 100% (4576/4576), 42.55 MiB | 19.59 MiB/s, done.
Resolving deltas: 100% (2823/2823), done.
root@test-redelk:/opt# cd RedELK/
root@test-redelk:/opt/RedELK# git fetch origin pull/152/head:pr152
remote: Enumerating objects: 17, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 17 (delta 11), reused 17 (delta 11), pack-reused 0
Unpacking objects: 100% (17/17), done.
From https://github.com/outflanknl/RedELK
 * [new ref]         refs/pull/152/head -> pr152
root@test-redelk:/opt/RedELK# git checkout pr152
Switched to branch 'pr152'
root@test-redelk:/opt/RedELK# echo master > VERSION
root@test-redelk:/opt/RedELK# sed -i 's/MODIFYME, for example //g' /opt/RedELK/certs/config.cnf
root@test-redelk:/opt/RedELK# vi certs/config.cnf
root@test-redelk:/opt/RedELK# ./initial-setup.sh ./certs/config.cnf



    ____            _  _____  _      _  __
   |  _ \  ___   __| || ____|| |    | |/ /
   | |_) |/ _ \ / _  ||  _|  | |    | ' /
   |  _ <|  __/| (_| || |___ | |___ | . \
   |_| \__\___| \____||_____||_____||_|\_\



This script will generate necessary keys and packages for RedELK deployments



[*] Will generate TLS certificates for the following DNS names and/or IP addresses:
DNS.1 = DOMAIN.TLD
IP.1 = XXX.XXX.XXX.XXX

[!] Make sure your ELK server will be reachable on these DNS names or IP addresses or your TLS setup will fail!
[*] Abort within 10 seconds to correct if needed.
[*] Creating certs dir if necessary
[*] Generating private key for CA
[*] Creating Certificate Authority
[*] Generating private key for ELK server
[*] Generating certificate for ELK server
[*] Signing certificate of ELK server with our new CA
[*] Converting ELK server private key to PKCS8 format
[*] Copying certificates to relevant redir and c2servers folders.
[*] Creating ssh directories if necessary
[*] Generating SSH key pair for scponly user
[*] Copying sshkeys to relevant folders.
[*] Copying VERSION file to subfolders.
[*] Creating TGZ packages for easy distribution


[*] Done with initial setup.
[*] Copy the redirs.tgz, c2servers.tgz and elkserver.tgz packages to every redirector, c2servers or ELK-server. Then run the relevant setup script there locally.

root@test-redelk:/opt/RedELK# cd elkserver/
root@test-redelk:/opt/RedELK/elkserver# vi mounts/redelk-config/etc/redelk/config.json
root@test-redelk:/opt/RedELK/elkserver# ./install-elkserver.sh



    ____            _  _____  _      _  __
   |  _ \  ___   __| || ____|| |    | |/ /
   | |_) |/ _ \ / _  ||  _|  | |    | ' /
   |  _ <|  __/| (_| || |___ | |___ | . \
   |_| \__\___| \____||_____||_____||_|\_\





This script will install and configure necessary components for RedELK on ELK server



[*] Memory found available for RedELK: 15333 MB.
[*] 15-16GB memory found


[*] No 'limited' parameter found. Going for the full RedELK installation including:
- RedELK
- Jupyter notebooks
- BloodHound / Neo4j

5 Seconds to abort

[*] Starting pre installation checks
[*] .env file doesn't exist yet, copying from .env.tmpl
[*] Setting RedELK version in docker env file
[*] Setting kibana_system ES password
[*] Setting logstash_system ES password
[*] Setting redelk_ingest ES password
[*] Setting redelk password in elasticsearch
[*] Setting redelk password in htaccess
[*] Setting elastic ES password in Docker template
[*] Setting elastic ES password in redelk config.json
[*] Setting Kibana encryption key
[*] Adjusting memory settings for ES
[*] Adjusting memory settings for NEO4J
[*] Setting neo4j password
[*] Setting permissions on logstash configs
[*] Setting permissions on redelk logs
[*] Setting permissions on Jupyter notebook working dir
[*] Validating configured external domain name DOMAIN.TLD
[*] Domain DOMAIN.TLD seems valid. Continuing.
[*] Setting external domain name in Docker env file
[*] Setting Let's Encrypt email in Docker env file
[*] Setting CERTS_DIR_NGINX_LOCAL
[*] Setting CERTS_DIR_NGINX_CA_LOCAL
[*] Setting TLS_NGINX_CRT_PATH
[*] Setting TLS_NGINX_KEY_PATH
[*] Setting TLS_NGINX_CA_PATH
[*] Adjusting Nginx config file
[*] Linking docker-compose.yml to the docker file used
[*] Creating password file for easy reference
[*] Copying password file for use with jupyter notebooks
[*] Running initial Let's Encrypt script
### Creating dummy certificate for DOMAIN.TLD ...
Creating network "redelk_net" with driver "bridge"
Creating volume "redelk_kibana_data" with local driver
Creating volume "redelk_bloodhound_data" with local driver
Creating volume "redelk_es_data" with local driver
Creating redelk-elasticsearch ...
Creating redelk-jupyter ...
Creating redelk-elasticsearch
Creating redelk-jupyter ... done
Creating redelk-bloodhound ...
Creating redelk-elasticsearch ... done
Creating redelk-logstash ...
Creating redelk-logstash ... done
Creating redelk-kibana ...
Creating redelk-kibana ... done
Creating redelk-nginx ...
Creating redelk-nginx ... done
Generating a RSA private key
..............................................................................................................++++
.......................................................................................................++++
writing new private key to '/etc/letsencrypt/live/DOMAIN.TLD/privkey.pem'
-----

### Starting nginx ...
Recreating redelk-jupyter ...
Recreating redelk-jupyter
Recreating redelk-elasticsearch ...
Recreating redelk-elasticsearch ... done
Recreating redelk-logstash ...
Recreating redelk-jupyter ... done
Recreating redelk-bloodhound ...
Recreating redelk-logstash ... done
Recreating redelk-kibana ...
Recreating redelk-kibana ... done
Recreating redelk-nginx ...
Recreating redelk-nginx ... done

### Requesting Let's Encrypt certificate for DOMAIN.TLD ...
### Removing dummy certificate folder
Starting redelk-jupyter ... done
Starting redelk-elasticsearch ... done
Starting redelk-bloodhound ...
Starting redelk-bloodhound ... done
Starting redelk-kibana ... done
Starting redelk-nginx ... done
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator webroot, Installer None
Account registered.
Requesting a certificate for DOMAIN.TLD
Performing the following challenges:
http-01 challenge for DOMAIN.TLD
Using the webroot path /var/www/certbot for all unmatched domains.
Waiting for verification...
Cleaning up challenges

IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at:
   /etc/letsencrypt/live/DOMAIN.TLD/fullchain.pem
   Your key file has been saved at:
   /etc/letsencrypt/live/DOMAIN.TLD/privkey.pem
   Your certificate will expire on 2021-07-27. To obtain a new or
   tweaked version of this certificate in the future, simply run
   certbot again. To non-interactively renew *all* of your
   certificates, run "certbot renew"
 - If you like Certbot, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le


### Reloading nginx ...
2021/04/28 20:57:23 [notice] 32#32: signal process started
[*] Building RedELK from redelk-full.yml file. Docker output below.

redelk-jupyter is up-to-date
redelk-elasticsearch is up-to-date
redelk-bloodhound is up-to-date
redelk-logstash is up-to-date
redelk-kibana is up-to-date
redelk-nginx is up-to-date
Creating redelk-base ...
Creating redelk-certbot ...
Creating redelk-base
Creating redelk-certbot ... done



 Done with base setup of RedELK on ELK server
 You can now login to the following interfaces:
   - Main RedELK Kibana interface on port 443 (user: redelk, pass:YksMKL-2XY8zbV90EjoaTSTWXo3kpT_q)
   - Jupyter notebooks on /jupyter (user: redelk, pass:YksMKL-2XY8zbV90EjoaTSTWXo3kpT_q)
   - Neo4J Browser port 7473 (user: neo4j, pass:9H35w9uCriUT0CCOyfgjTuME83BafnW4)
   - Neo4J using the BloodHound app on port 7687 (user: neo4j, pass:9H35w9uCriUT0CCOyfgjTuME83BafnW4)



 !!! WARNING
 !!! WARNING - IF YOU WANT FULL FUNCTIONALITY YOU ARE NOT DONE YET !!!
 !!! WARNING

 You should really:
   - adjust the mounts/redelk-config/etc/cron.d/redelk file to include your teamservers
   - adjust all config files in mounts/redelk-config/etc/redelk to include your specifics like VT API, email server details, etc
   - adjust the .env file to match any specifics you need (e.g. using custom certificate, etc.)
```

**Comment by drfinkus** [comment]
@fastlorenzo tested now. Can confirm #152 now works for the full install, which is great.

However, for the limited install (i.e. `./install-elkserver.sh limited`) I still run into the same `connection refused` error as before for certbot.

**Comment by fastlorenzo** [comment]
@drfinkus I've done exactly the same steps, and used `./install-elkserver.sh limited`, and got a valid certificate (using #152)
@MarcOverIP I propose to merge this one as it's been tested with the following use cases:
- With LE, full install: working
- With LE, limited install: working
- Without LE, full install: working
- Without LE, limited install: working

**Comment by MarcOverIP** [comment]
Merged #152

@drfinkus could you let us know if you still experience issues?

**Comment by drfinkus** [comment]
@fastlorenzo @MarcOverIP tested the `master` branch now on a fresh 18.04, can confirm the issue is solved both with full and limited install. Thanks a lot for all the support fixing this! I'll go ahead and close this if that's okay with you, and will do the same with the other, related issue.
